# Original file:

In [1]:
#!/usr/bin/python

""" 
    This is the code to accompany the Lesson 2 (SVM) mini-project.

    Use a SVM to identify emails from the Enron corpus by their authors:    
    Sara has label 0
    Chris has label 1
"""
    
import sys
from time import time
sys.path.append("../tools/")
from email_preprocess import preprocess


### features_train and features_test are the features for the training
### and testing datasets, respectively
### labels_train and labels_test are the corresponding item labels
features_train, features_test, labels_train, labels_test = preprocess()




#########################################################
### your code goes here ###
from sklearn.svm import SVC

#########################################################





/Users/rexwang/anaconda3/envs/udacity_machine_learning_py27/lib/python2.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


no. of Chris training emails: 7936
no. of Sara training emails: 7884


# linear svm accuracy and time

In [8]:
clf = SVC(kernel="linear")
t0 = time()
clf.fit(features_train, labels_train)
print "Training time:", round(time()-t0,3), "s"

t0 = time()
score = clf.score(features_test, labels_test)
print "Predict time:", round(time()-t0,3), "s"
print("Accuracy: %.3f" % score)

Training time: 144.706 s
Predict time: 14.735 s
Accuracy: 0.984


# Smaller training set

In [9]:
small_features_train = features_train[:len(features_train)/100] 
small_labels_train = labels_train[:len(labels_train)/100]
clf = SVC(kernel="linear")
t0 = time()
clf.fit(small_features_train, small_labels_train)
print "Training time:", round(time()-t0,3), "s"

t0 = time()
score = clf.score(features_test, labels_test)
print "Predict time:", round(time()-t0,3), "s"
print("Accuracy: %.3f" % score)

Training time: 0.085 s
Predict time: 0.887 s
Accuracy: 0.885


# RBF kernel with small training set

In [10]:
clf = SVC()
t0 = time()
clf.fit(small_features_train, small_labels_train)
print "Training time:", round(time()-t0,3), "s"

t0 = time()
score = clf.score(features_test, labels_test)
print "Predict time:", round(time()-t0,3), "s"
print("Accuracy: %.3f" % score)

Training time: 0.099 s
Predict time: 1.005 s
Accuracy: 0.616


# RBF with different C

In [12]:
for i in [10, 100, 1000, 10000, 100000]:
    print("C = %d" % i)
    clf = SVC(C=i)
    t0 = time()
    clf.fit(small_features_train, small_labels_train)
    print "  Training time:", round(time()-t0,3), "s"

    t0 = time()
    score = clf.score(features_test, labels_test)
    print "  Predict time:", round(time()-t0,3), "s"
    print("  Accuracy: %.3f" % score)

C = 10
  Training time: 0.094 s
  Predict time: 0.997 s
  Accuracy: 0.616
C = 100
  Training time: 0.089 s
  Predict time: 1.02 s
  Accuracy: 0.616
C = 1000
  Training time: 0.095 s
  Predict time: 0.966 s
  Accuracy: 0.821
C = 10000
  Training time: 0.088 s
  Predict time: 0.786 s
  Accuracy: 0.892
C = 100000
  Training time: 0.088 s
  Predict time: 0.72 s
  Accuracy: 0.860


C = 10000 gives the best accuracy (0.892)

# RBF with full training data

In [14]:
clf = SVC(C=10000)
t0 = time()
clf.fit(features_train, labels_train)
print "Training time:", round(time()-t0,3), "s"

t0 = time()
score = clf.score(features_test, labels_test)
print "Predict time:", round(time()-t0,3), "s"
print("Accuracy: %.3f" % score)

Training time: 107.333 s
Predict time: 10.035 s
Accuracy: 0.991


# Get prediction result for test 10, 26 and 50

In [16]:
pred = clf.predict(features_test)
print("Predict result [10] = %d, [26] = %d, [50] = %d" % (pred[10],pred[26],pred[50]))

Predict result [10] = 1, [26] = 0, [50] = 1


# Find total number of predictions for class 1 

In [22]:
import numpy as np
print("Number of predictions for class 1: %d" % np.count_nonzero(pred))

Number of predictions for class 1: 877
